### PR TITLE
[WOR-492] Check that a user is enabled before getting a workspace context

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
[Relevant ticket](https://broadworkbench.atlassian.net/browse/WOR-492)

We are varying the error messaging for disabled users when fetching a workspace depending on whether the workspace exists.

## This PR
* Adds a check with SAM to determine if the user is disabled when getting a workspace. If so, the get() call fails out with the same error message, regardless of whether the workspace exists. 

## TODO
- [x] Update tests

---------
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala]: It feels safer to me to have this default to false. How painful would the refactor be to have this default to false instead of true? Maybe the `apply` in the object sets `enabled=true` if that's only being used in tests

Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala]: Yeah, this is a good point. I think tacking on `enabled` to RawlsUser when it's possible that there are call sites that will not know the enabled/disabled status is a code smell. I've introduced a new `SamUserStatusResponse` model in this [commit](https://github.com/broadinstitute/rawls/pull/2025/commits/aea707f038a4e68a4233fb527885d64e40aae91a) that will avoid this confusion.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/mock/MockSamDAO.scala
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/mock/MockSamDAO.scala:16]: Re-using `RawlsUser` as the return type was forcing me to tack on an `enabled` val to that case class, with a default of `true` (see @marctalbott 's point earlier in the PR). Since there are only two call sites for this method, shifting to a first-class SAM model seems semantically clearer.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala:35]: This method and the related call on L189 are the nut of the fix.

**Review by cahrens — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/SamModel.scala
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/model/SamModel.scala]: nit: unused import

**Review by marctalbott — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala]: optional, I wouldn't block the PR on this … if you felt like consolidating:
```scala
      case Some(user) if (user.enabled) => Future.successful()
      case _ => Future.failed(new UserDisabledException(StatusCodes.Unauthorized, "Unauthorized"))
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala:187]: thinking out loud: I poked at this a bit to find a good way to avoid always making this Sam call. The best I could come up with was adding a recover/catch and only making this extra Sam call on an error, which would eliminate the extra Sam call from the typical/successful path. The code gets more complex with that. Again, not gonna block the PR on this, not requesting changes, just talking.

**Review by davidangb — APPROVED** [review]
one nit and two TOLs inline, giving this a :+1: and your call on whether or not to change anything based on my feedback.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala:187]: Yeah, that's a good point. I'm inclined to favor a simpler/more obvious codepath for now; we can optimize down the line if need be.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala]: Good call, implemented.

**Review by davidangb — APPROVED** [review]

**Pull request merged**
